### PR TITLE
Hdi ha services

### PR DESCRIPTION
**Pull request by jieyan1269**

### CODE REVIEW

File: ClusterManagement/hdinsight-ha-services.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+•	Livy for Spark2 Server
+
+HDInsight utilizes HDInsight zookeeper to determine status of HA services and does failover accordingly. The HDI zookeeper is another quorum of zookeeper server running on zookeeper nodes in parallel with Apache zookeeper, which is used to decide the active headnode. The HDI HA services only run on headnodes. They should always be running only on the active headnode and should be stopped and put in maintenance mode on the standby headnode.
+
+### **Startup**
+During deployment, HDInsight agent starts HA-service-related components in the order of HDI zookeeper, master failover controller, and slave failover controller.
+
+### **Slave failover controller**
+
+The slave failover controller runs on every node of the cluster. It is responsible for starting Ambari agent and slave-ha-service, which is an HA service handler, on each node. It also periodically queries HDI zookeeper about the active headnode. If there is a change in the active/standby headnodes, the slave failover controller updates the hosts configuration file, restarts Ambari agent and slave-ha-service. The slave-ha-service is responsible for stopping HDI HA services on standby headnode.
```
[the remark this answers is not in the document]
Comment (jieyan1269) [reply ClusterManagement/hdinsight-ha-services.md]: slave-ha-service on standby headnode should stop ha services when failover happens, but there are some exceptions showing in the log (we may need to investigate and fix it):

2019-08-27 01:00:55,943 [myid:] - INFO  [main:HAServiceHandler@?] - Stopping HA services in host hn0-hadoop.vbk3buvahw0e5hiep4fuxaze1c.cx.internal.cloudapp.net
2019-08-27 01:00:56,003 [myid:] - INFO  [main:HAServiceHandler@?] - HostComponent [component_name=APP_TIMELINE_SERVER, desired_state=INSTALLED, host_name=hn0-hadoop.vbk3buvahw0e5hiep4fuxaze1c.cx.internal.cloudapp.net, state=INSTALLED]
2019-08-27 01:00:56,004 [myid:] - INFO  [main:HAServiceHandler@?] - HostComponent [component_name=HISTORYSERVER, desired_state=STARTED, host_name=hn0-hadoop.vbk3buvahw0e5hiep4fuxaze1c.cx.internal.cloudapp.net, state=STARTED]
2019-08-27 01:00:56,023 [myid:] - ERROR [main:HAServiceHandler@?] - java.io.EOFException: No content to map to Object due to end of input
2019-08-27 01:00:56,024 [myid:] - WARN  [main:RetryUtil@?] - Operation failed with exception 'java.io.EOFException: No content to map to Object due to end of input' remaining retry count: 2

File: ClusterManagement/hdinsight-ha-services.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+### **The failover process**
+There is a health monitor, which is a daemon, running along with each master failover controller to periodically heartbeats with the headnodes. The headnode is regarded as an HA service in this scenario. The health monitor checks if the HA service is healthy and if it is ready to join in the leadership election, if yes, this HA service will compete in the election, otherwise, it will quit the election until it becomes ready again. In the case of active headnode failures, such as headnode crash, unresponsive, rebooting, etc., if the standby headnode achieves the leadership and becomes active, its master failover controller will start all HDI HA services on it and stop these services on the other headnode. In case of HDI HA service failures, such as service down, unhealthy, etc., master failover controller should be able to automatically restart or stop the services according to the headnode status. Users should not manually start HDI HA services on both head nodes, instead, allow automatic or manual failover to recover the problem.
+
+### **HDFS NameNode high availability**
```
[the remark this answers is not in the document]
Comment (jieyan1269) [reply ClusterManagement/hdinsight-ha-services.md]: Done

File: ClusterManagement/hdinsight-ha-services.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+HDInsight clusters based on Hadoop 2.0 or higher also provide NameNode high availability. There are two namenodes running on two headnodes, respectively, which are configured for automatic failover. The namenodes utilize ZKFailoverController to communicate with Apache zookeeper to elect for active/standby status. ZKFailoverController runs on both headnodes and works in the same way as the master failover controller above. Please notices that Apache zookeeper is independent of HDI zookeeper, thus, active namenode may not run on active headnode. When the active namenode is dead or unhealthy, the standby namenode will win the election and become active.
+
+### **Gaps**
```
[the remark this answers is not in the document]
Comment (jieyan1269) [reply ClusterManagement/hdinsight-ha-services.md]: Done

File: ClusterManagement/hdinsight-ha-services.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (csunilkumar) [reply ClusterManagement/hdinsight-ha-services.md]: @jieyan1269 are waiting on anyone on this for a pull request

**Comment by csunilkumar** [comment]
@vijaysr this is waiting on your for review,

**Pull request merged**